hw/ssi/spi_gpio.c

```diff
@@ -25,22 +25,89 @@
 #include "hw/ssi/spi_gpio.h"
 #include "hw/irq.h"
 
-static void cs_handler(void *opaque, int n, int level)
+#define SPI_CPHA	BIT(0)	/* clock phase (1 = SPI_CLOCK_PHASE_SECOND) */
+#define SPI_CPOL	BIT(1)	/* clock polarity (1 = SPI_POLARITY_HIGH) */
+
+static void do_leading_edge(SpiGpioState *s)
+{
+    int cpha = !!(s->mode & SPI_CPHA);
+
+    if (cpha) {
+       s->output_byte |= (s->mosi << s->clk_counter);
+    } else {
+       s->input_byte <<= 1;
+       s->input_byte |= s->mosi ? 1 : 0;
+    }
+}
+
+static void do_trailing_edge(SpiGpioState *s)
+{
+    int cpha = !!(s->mode & SPI_CPHA);
+
+    if (cpha) {
+        s->output_byte |= (s->mosi << s->clk_counter);
+    } else {
+        s->miso = !!(s->output_byte & 0x80);
+        s->miso <<= 1;
+        object_property_set_bool(OBJECT(s->aspeed_gpio),
+                                "gpioX5", s->miso, NULL);
+    }
+}
+
+static void do_ssi_transfer(SpiGpioState *s)
+{
+    s->output_byte = ssi_transfer(s->spi, s->input_byte);
+}
+
+static void cs_set_level(void *opaque, int n, int level)
 {
     SpiGpioState *s = SPI_GPIO(opaque);
     s->cs = !!level;
 
     /* relay the CS value to the CS output pin */
     qemu_set_irq(s->cs_output_pin, s->cs);
+
+    s->miso = !!(s->output_byte & 0x80);
+    s->miso <<= 1;
+    object_property_set_bool(OBJECT(s->aspeed_gpio),
+                            "gpioX5", s->miso, NULL);
+
+    s->clk = !!(s->mode & SPI_CPOL);
 }
 
-static void clk_handler(void *opaque, int n, int level)
+static void clk_set_level(void *opaque, int n, int level)
 {
     SpiGpioState *s = SPI_GPIO(opaque);
+
+    /* since clk_counter begins at 0 */
+    if (s->clk_counter == 8) {
+        do_ssi_transfer(s);
+        s->clk_counter = 0;
+        s->input_byte = 0;
+    }
+
+    bool pre = s->clk;
     s->clk = !!level;
+    bool cur = s->clk;
+
+    /* CS# is high/not selected, do nothing */
+    if (s->cs) {
+        return;
+    }
+
+   /* leading edge */
+    if (!pre && cur) {
```

Comment:
if we based on leading and tailing edge
```suggestion
    if ( (pre==s->cidle) && (cur != s->cidle)) {
```